### ISSUE TITLE
torch.where() can produce nan values for unselected branch during backward

### ISSUE THREAD

**Issue by lfrommelt** [pytorch/pytorch#156212]
### 🐛 Describe the bug

Thanks in advance. Since the not selected branch never contributes to the loss, I was not expecting it to produce gradients. This was terrible to debug, rather uncommon use-case

## Minimal code for replicating bug:
```python
# some arbitrary max value for sth similar to torch.clamp(x, max=x_th)
x_th=torch.tensor(10)

# nan value, just for this experiment
x = torch.tensor([1.0, float('nan'), 1e10], requires_grad=True)

# some arbitrary secondary path, for showcase
b = torch.tensor(1)

# this never selects nan and protects against overflows
x_out = torch.where(x <= x_th, torch.exp(x), b)

print(x_out)

l = x_out.sum()
l.backward()

print(x.grad)
```

## Expected behaviour:
```python
# some arbitrary max value for sth similar to torch.clamp(x, max=x_th)
x_th=torch.tensor(10)

# nan value, just for this experiment
x = torch.tensor([1.0, float('nan'), 1e10], requires_grad=True)

# some arbitrary secondary path, for showcase
b = torch.tensor(1)

# this never selects nan and protects against overflows
x_out=torch.ones_like(x)
x_out[x<=x_th]=torch.exp(x[x<=x_th])
x_out[~(x<x_th)]=b

print(x_out)

l = x_out.sum()
l.backward()

print(x.grad)
```

## Side note
If anything, I would have expected the gradient of exp(1e10) to be inf, not nan. This looks a bit, like the forward of `where` actually computes both branches and then simply multiplies the unselected one with zero, instead of masking (during backward, gradient of 0*inf).

### Versions

Collecting environment information...
PyTorch version: 2.7.0+cpu
Is debug build: False
CUDA used to build PyTorch: Could not collect
ROCM used to build PyTorch: N/A

OS: Microsoft Windows 11 Pro (10.0.26100 64-bit)
GCC version: Could not collect
Clang version: Could not collect
CMake version: Could not collect
Libc version: N/A

Python version: 3.12.8 | packaged by Anaconda, Inc. | (main, Dec 11 2024, 16:48:34) [MSC v.1929 64 bit (AMD64)] (64-bit runtime)
Python platform: Windows-11-10.0.26100-SP0
Is CUDA available: False
CUDA runtime version: 12.6.85
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: GPU 0: NVIDIA RTX A500 Laptop GPU
Nvidia driver version: 572.40
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Name: 13th Gen Intel(R) Core(TM) i7-1370P
Manufacturer: GenuineIntel
Family: 198
Architecture: 9
ProcessorType: 3
DeviceID: CPU0
CurrentClockSpeed: 1492
MaxClockSpeed: 1900
L2CacheSize: 11776
L2CacheSpeed: None
Revision: None

Versions of relevant libraries:
[pip3] numpy==2.2.5
[pip3] torch==2.7.0
[conda] numpy                     2.2.5                    pypi_0    pypi
[conda] torch                     2.7.0                    pypi_0    pypi

**Comment by lfrommelt** [comment]
I just realized that this is a duplicate from #108618 and #68425. Sorry for that. Sorry for that. Anyways, could you please fix/change this, or add a warning to the documentation?

**Comment by ngimel** [comment]
As discussed in the linked issues, it won't be fixed, there's no reasonable way to fix that.
